src/cubicbez.rs

```diff
@@ -73,6 +76,197 @@ impl CubicBez {
+    }
+
+    // Approximate a cubic curve with a quadratic spline of `n` curves
+    fn approx_spline_n(&self, n: usize, accuracy: f64) -> Option<QuadSpline> {
+        if n == 1 {
+            return self
+                .try_approx_quadratic(accuracy)
+                .map(|quad| QuadSpline::new(vec![quad.p0, quad.p1, quad.p2]));
+        }
+        let mut cubics = self.split_into_n(n);
+
+        // The above function guarantees that the iterator returns n items,
+        // which is why we're unwrapping things with wild abandon.
+        let mut next_cubic = cubics.next().unwrap();
+        let mut next_q1: Point = next_cubic.approx_quad_control(0.0);
+        let mut q2 = self.p0;
+        let mut d1 = Vec2::ZERO;
+        let mut spline = vec![self.p0, next_q1];
+        for i in 1..=n {
+            let current_cubic: CubicBez = next_cubic;
+            let q0 = q2;
+            let q1 = next_q1;
+            q2 = if i < n {
+                next_cubic = cubics.next().unwrap();
+                next_q1 = next_cubic.approx_quad_control(i as f64 / (n - 1) as f64);
+
+                spline.push(next_q1);
+                q1.midpoint(next_q1)
+            } else {
+                current_cubic.p3
+            };
+            let d0 = d1;
+            d1 = q2.to_vec2() - current_cubic.p3.to_vec2();
+
+            if d1.hypot() > accuracy
+                || !CubicBez::new(
+                    d0.to_point(),
+                    q0.lerp(q1, 2.0 / 3.0) - current_cubic.p1.to_vec2(),
+                    q2.lerp(q1, 2.0 / 3.0) - current_cubic.p2.to_vec2(),
+                    d1.to_point(),
+                )
+                .fit_inside(accuracy)
+            {
+                return None;
+            }
+        }
+        spline.push(self.p3);
+        Some(QuadSpline::new(spline))
+    }
+
+    fn approx_quad_control(&self, t: f64) -> Point {
+        let p1 = self.p0 + (self.p1 - self.p0) * 1.5;
+        let p2 = self.p3 + (self.p2 - self.p3) * 1.5;
+        p1.lerp(p2, t)
+    }
+
+    /// Approximate a cubic with a single quadratic
+    ///
+    /// Returns a quadratic approximating the given cubic that maintains
+    /// endpoint tangents if that is within tolerance, or None otherwise.
+    fn try_approx_quadratic(&self, accuracy: f64) -> Option<QuadBez> {
+        if let Some(q1) = Line::new(self.p0, self.p1).crossing_point(Line::new(self.p2, self.p3)) {
+            let c1 = self.p0.lerp(q1, 2.0 / 3.0);
+            let c2 = self.p3.lerp(q1, 2.0 / 3.0);
+            if !CubicBez::new(
+                Point::ZERO,
+                c1 - self.p1.to_vec2(),
+                c2 - self.p2.to_vec2(),
+                Point::ZERO,
+            )
+            .fit_inside(accuracy)
+            {
+                return None;
+            }
+            return Some(QuadBez::new(self.p0, q1, self.p3));
+        }
+        None
+    }
+
+    fn split_into_n(&self, n: usize) -> impl Iterator<Item = CubicBez> {
```

Comment:
out of curiosity: do we have a sense of the distribution of `n` in calls to this function? If it is more likely to be a small value than a large value it might be worth getting a little fancy in order to avoid the allocation.

Looking through this a bit more, I think it should be fairly trivial to avoid the allocation, if so desired, and convert this whole method to use `std::iter::from_fn`?

Reply:
As it happens, yes: https://github.com/googlefonts/cu2qu/blob/6812ae48c6b81aee95949d64e2a9489c333d825d/tests/cu2qu_test.py#L70-L77

```python
        expected = {
            2: 6,
            3: 26,
            4: 82,
            5: 232,
            6: 360,
            7: 266,
            8: 28}
```

Reply:
okay lemme take a poke at this and see if I can get rid of the allocation.

Reply:
Okay I got a bit nerd-sniped here. There's a non-allocating version at https://gist.github.com/cmyr/1d6c316bec50c4d23e4be2c794fe6b89, feel free to take it if you like?